pylops/optimization/leastsquares.py

```diff
@@ -106,65 +107,66 @@ def NormalEquationsInversion(
     implicitly assumed to be zero.
 
     """
-    ncp = get_array_module(data)
+    with disabled_ndarray_multiplication():
```

Comment:
Why is this needed here? And not in other solvers in this module?

Reply:
I was having an issue with this solver before because it calls the SciPy cg which does not accept NDarray multiplication. However, I looked into it a bit deeper and it seems that there are several instances in the codebase where methods that only work for 1d arrays are called (e.g. scipy lsqr and cg). They are mostly internal, and since out tests are all built for v1.x behavior, I didn't catch this.

One solution is to create a wrapper for these scipy functions so that users can understand how to interact with functions that might be reliant on v1.x behavior (but that they cannot change/are external). It would work something like this:

```python
# v1.x code: only works with 1d arrays
from scipy.sparse.linalg import lsqr
xlsqr, *_ = lsqr(Sop, y.ravel())

# proposed new code
def add_ndarray_support_scipy(func): # this function would be in pylops.utils.decorators
    """
    Decorator which converts a SciPy-solver-type 
    function into one that supports ndarray multiplication.
    """
    def wrapper(A, b, *args, **kwargs): # this is the signature of sparse solvers in scipy
        with disabled_ndarray_multiplication():
            res = func(A, b.ravel(), *args, **kwargs)
            res[0] = res[0].reshape(A.dims)
            return res

from scipy.sparse.linalg import lsqr
lsqr = add_ndarray_support_scipy(lsqr) # "monkey-patch"
xlsqr, *_ = lsqr(Sop, y) # works with ndarrays

```

Reply:
Mmh, I see. Yes we have quite a few functions that call a solver, and if the input is a numpy array by default they would call the appropriate scipy solver  (already if it is cupy they will call our equivalent as at that time cupy did not have scipy solvers in place for linear operators.. I need to check if they do now). 

We could to begin with do this, and in the longer run when all our solvers are refactored change them so that they can handle multiple rhs, then check if this is the case and use them instead of scipy ones… or even switch all together to our solvers and just have some tutorial that shows you can use scipy solvers but we now reccomend using ours?

Reply:
I implemented the above method (with some small alterations) and in general it works. I also added in the migration readme how to use operators with the new LinearOperator (either disable nd-array mult or decorate them).

The only thing missing are the sparse solvers, but I need to look into those a bit better. For now the solution to those is to disable nd-array support when calling them.